src/site/content/en/blog/2d-graphics-in-emscripten/index.md

```diff
@@ -0,0 +1,510 @@
+
+#### Unblocking event loop with "main loop" APIs
+
+[`emscripten_set_main_loop`](https://emscripten.org/docs/api_reference/emscripten.h.html#c.emscripten_set_main_loop) doesn't require any compiler transformations for unwinding and rewinding the call stack, and that way avoids the code size overhead. However, in exchange, it requires a lot more manual modifications to the code.
+
+First, the body of the event loop needs to be extracted into a separate function. Then, `emscripten_set_main_loop` needs to be called with that function as a callback in the first argument, an FPS in the second argument (`0` for the native refresh interval), and a boolean indicating whether to simulate infinite loop (`true`) in the third:
+
+```cpp
+emscripten_set_main_loop(callback, 0, true);
+```
+
+The newly created callback won't have any access to the stack variables in the `main` function, so variables like `window` and `renderer` need to be either extracted into a heap-allocated struct and its pointer passed via [`emscripten_set_main_loop_arg`](https://emscripten.org/docs/api_reference/emscripten.h.html#c.emscripten_set_main_loop_arg) variant of the API, or extracted into global `static` variables (I went with the latter for simplicity). The result is slightly harder to follow, but it draws the same rectangle as the last example:
+
+```cpp/20
+#include <SDL2/SDL.h>
+#include <stdio.h>
+#ifdef __EMSCRIPTEN__
+#include <emscripten.h>
+#endif
+
+SDL_Window *window;
+SDL_Renderer *renderer;
+
+bool handle_events() {
+  SDL_Event event;
+  SDL_PollEvent(&event);
+  if (event.type == SDL_QUIT) {
+    return false;
+  }
+  return true;
+}
+
+void run_main_loop() {
+#ifdef __EMSCRIPTEN__
+  emscripten_set_main_loop([]() { handle_events(); }, 0, true);
+#else
+  while (handle_events())
+    ;
+#endif
+}
+
+int main() {
+  SDL_Init(SDL_INIT_VIDEO);
+
+  SDL_CreateWindowAndRenderer(300, 300, 0, &window, &renderer);
+
+  SDL_SetRenderDrawColor(renderer, /* RGBA: green */ 0x00, 0x80, 0x00, 0xFF);
+  SDL_Rect rect = {.x = 10, .y = 10, .w = 150, .h = 100};
+  SDL_RenderFillRect(renderer, &rect);
+  SDL_RenderPresent(renderer);
+
+  run_main_loop();
+
+  SDL_DestroyRenderer(renderer);
+  SDL_DestroyWindow(window);
+
+  SDL_Quit();
+}
+```
+
+Since all the control flow changes are manual and reflected in the source code, it can be compiled without the Asyncify feature again:
+
+```bash
+emcc example.cpp -o example.html -s USE_SDL=2
+```
+
+This example might seem useless, because it works no differently from the first version, where the rectangle was drawn on canvas successfully despite the code being a lot simpler, and the `SDL_QUIT` event - the only one handled in the `handle_events` function - is ignored on the web anyway.
```

Comment:
```suggestion
This example might seem useless, because it works no differently from the first version, where the rectangle was drawn on canvas successfully despite the code being a lot simpler, and the `SDL_QUIT` event&mdash;the only one handled in the `handle_events` function&mdash;is ignored on the web anyway.
```